augur/index.py

```diff
@@ -206,9 +206,8 @@ def run(args):
             tot_length = None
         else:
             num_of_seqs, tot_length = index_sequences(args.sequences, args.output)
-    except ValueError as error:
-        print("ERROR: Problem reading in {}:".format(sequences_path), file=sys.stderr)
-        print(error, file=sys.stderr)
+    except FileNotFoundError:
+        print(f"ERROR: Could not open sequences file '{args.sequences}'.", file=sys.stderr)
         return 1
```

Comment:
```suggestion
    except FileNotFoundError:
        print(f"ERROR: Could not open sequences file '{args.sequences}'.", file=sys.stderr)
        return 1
    except ValueError as error:
        print(f"ERROR: Problem reading sequences file '{args.sequences}':", file=sys.stderr)
        print(error, file=sys.stderr)
        return 1
```

It looks like the original intent was to catch and print with an additional line, something like:

```
ERROR: Problem reading in ...
character set ... contains a non-string element: ...
```

Since `index_sequences() -> index_sequence()` raises custom `ValueErrors`:

https://github.com/nextstrain/augur/blob/ec76046d4989df168a626aac7a26efcdcacdcd10/augur/index.py#L121-L132

If `ValueError` isn't caught here, it will result in an uncaught exception.

Reply:
That's a good point, @victorlin. This is a weird situation, since those value errors can only occur when the `values` passed by `index_sequences` to `index_sequence` are invalid. There is no way to test that scenario as a functional test, since the user cannot override those values. If the `ValueError` gets raised, it's not a problem with the sequence input; it's a programming error (more akin to a syntax error).

We have [a general problem with Augur where we don't always catch exceptions and print helpful messages for the user](https://github.com/nextstrain/augur/issues/642). We don't have a general approach to handling unexpected exceptions like this, though.

What would you think about general catching any unexpected exceptions and reporting more details to the user (like the traceback), so they can report the issue to us?

```python
except Exception as error:
    print(
        f"There was an unexpected error: {error}",
        "Please consider reporting this error and the traceback included below as an issue at:",
        "https://github.com/nextstrain/augur/issues/new",
        file=sys.stderr
    )
    raise
```

If we took this approach, this sort of logic would probably live in [`augur.__main__`](https://github.com/nextstrain/augur/blob/002c44f94459ee5bf9111e5ece659c9e429fb1e5/augur/__main__.py#L9-L10) where it could provide the same interface across all subcommands.

Reply:
Thanks for the explanation! Yeah, a global exception handler sounds like a good idea regardless. But I see your point about this being a programming error so better to leave your change as-is.

Reply:
P.S. I built [a global exception handler into augur](https://github.com/nextstrain/augur/blob/master/augur/__init__.py#L74-L78), it's just that it only handles `RecursionError` at the moment!

Reply:
Thank you, @jameshadfield! I used this to create [an issue for this topic](https://github.com/nextstrain/augur/issues/903).